types/node/util.d.ts

```diff
@@ -1105,6 +1105,197 @@ declare module 'util' {
+     *     - `short` A single character alias for the option.
+     *
+     *   - `strict`: Whether an error should be thrown when unknown arguments
+     *     are encountered, or when arguments are passed that do not match the
+     *     `type` configured in `options`. **Default:** `true`.
+     *   - `allowPositionals`: Whether this command accepts positional arguments.
+     *     **Default:** `false` if `strict` is `true`, otherwise `true`.
+     *   - `tokens`: Whether tokens {boolean} Return the parsed tokens. This is useful
+     *     for extending the built-in behavior, from adding additional checks through
+     *     to reprocessing the tokens in different ways.
+     *     **Default:** `false`.
+     *
+     * @returns The parsed command line arguments:
+     *
+     *   - `values` A mapping of parsed option names with their string
+     *     or boolean values.
+     *   - `positionals` Positional arguments.
+     *   - `tokens` Detailed parse information (only if `tokens` was specified).
+     *
+     */
+    export function parseArgs<T extends ParseArgsConfig>(config: T): ParsedResults<T>;
+
+    interface ParseArgsOptionConfig {
+        type: 'string' | 'boolean';
+        short?: string;
+        multiple?: boolean;
+    }
+
+    interface ParseArgsOptionsConfig {
+        [longOption: string]: ParseArgsOptionConfig;
+    }
+
+    interface ParseArgsConfig {
+        strict?: boolean;
+        allowPositionals?: boolean;
+        tokens?: boolean;
+        options?: ParseArgsOptionsConfig;
+        args?: string[];
+    }
+
+    /*
+    IfDefaultsTrue and IfDefaultsFalse are helpers to handle default values for missing boolean properties.
+    TypeScript does not have exact types for objects: https://github.com/microsoft/TypeScript/issues/12936
+    This means it is impossible to distinguish between "field X is definitely not present" and "field X may or may not be present".
+    But we expect users to generally provide their config inline or `as const`, which means TS will always know whether a given field is present.
+    So this helper treats "not definitely present" (i.e., not `extends boolean`) as being "definitely not present", i.e. it should have its default value.
+    This is technically incorrect but is a much nicer UX for the common case.
+    The IfDefaultsTrue version is for things which default to true; the IfDefaultsFalse version is for things which default to false.
+    They differ only in the final item of the if-else chain.
+    */
+    type IfDefaultsTrue<T, IfTrue, IfFalse> = T extends true
+        ? IfTrue
+        : T extends false
+        ? IfFalse
+        : T extends boolean
+        ? IfTrue | IfFalse
+        : IfTrue;
+
+    type IfDefaultsFalse<T, IfTrue, IfFalse> = T extends true
+        ? IfTrue
+        : T extends false
+        ? IfFalse
+        : T extends boolean
+        ? IfTrue | IfFalse
+        : IfFalse;
```

Comment:
The `T extends boolean` condition isn't needed here as well.

Also, I would swap the first two conditions since `undefined` on its own compares like `any` when `strictNullChecks: false`. If the conditions aren't swapped, then `IfDefaultsFalse<undefined, A, B>` will result in `A` when not in strict mode.

Reply:
> The `T extends boolean` condition isn't needed here as well.

I think is is? If the option is omitted entirely, the type is `undefined`, which does not extend boolean. This is how the "defaults to false" condition is achieved.

That is, this is supposed to handle the following four cases:
- `T` is known to be `true`, in which case this should be `IfTrue`
- `T` is known to be `false`, in which case this should be `false`
- The option in question wasn't provided at all, in which case this should be `IfFalse` (hence "DefaultsFalse")
- The option in question _was_ provided but it can only be inferred to be a boolean, not a precise value, in which case this should be `IfTrue | IfFalse`.

Or maybe I'm misunderstanding; can you write out what you're suggesting in full?

> Also, I would swap the first two conditions since `undefined` on its own compares like any when `strictNullChecks: false`

Good catch, will do.

Reply:
> > The `T extends boolean` condition isn't needed here as well.
> 
> I think is is? If the option is omitted entirely, the type is `undefined`, which does not extend boolean. This is how the "defaults to false" condition is achieved.
> 
> That is, this is supposed to handle the following four cases (for an optional boolean config option):
> 
> * `T` is known to be `true`, in which case this should be `IfTrue`
> * `T` is known to be `false`, in which case this should be `IfFalse`
> * The option in question wasn't provided at all, in which case this should be `IfFalse` (hence "DefaultsFalse")
> * The option in question _was_ provided but it can only be inferred to be a boolean, not a precise value, in which case this should be `IfTrue | IfFalse`.

The last bullet point doesn't matter. Since the conditional type is distributive, that branch will never be reached. Consider the following:

```ts
type X<T> =
  T extends true ? 'true' :
  T extends false ? 'false' :
  T extends boolean ? 'boolean' :
  'default';

type A = X<true>; // 'true'
type B = X<false>; // 'false'
type C = X<boolean>; // 'true' | 'false'
type D = X<undefined>; // 'default'
```

[TypeScript Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAGgPAFQHxQLwCgpQVCAPYCAOwBMBnKYAJwFdoB+KAcmrqagC5NtcDjyoAMwCGAGzINmI8RHZcsOfIVIUARgHt1oiMKJRGTDVp1E53JiQgiao4EwDc6dKEhQAgmlhxWEJPagA9AHMPkzO4NAAQp7w0hJ+gcFMcbLhrgDCMXBG2roJQSG0slAAPlJiEmEu0AAiWTSkVgCWRBAk+UmW1rZMQA)

As you can see, for `C` the `'boolean'` result is never reached. When the checker encounters a naked type variable in the check type of the conditional (in this case, `T`) and that type is a union, the condition is distributed over the union re-evaluating the conditional type for each constituent. Since `boolean` is a union of `true | false`, the conditional type will match `true extends true ? 'true' : ...` for one constituent, and `false extends false ? 'false' : ...` for the other, and never need to descend any deeper.

> Or maybe I'm misunderstanding; can you write out what you're suggesting in full?

```ts
type IfDefaultsTrue<T, IfTrue, IfFalse> =
    T extends true ? IfTrue :
    T extends false ? IfFalse :
    IfTrue;

// IfDefaultsTrue<true, A, B>               -> A
// IfDefaultsTrue<undefined, A, B>          -> A
// IfDefaultsTrue<false, A, B>              -> B
// IfDefaultsTrue<boolean, A, B>            -> A | B

type IfDefaultsFalse<T, IfTrue, IfFalse> =
    T extends false ? IfFalse :
    T extends true ? IfTrue :
    IfFalse;

// IfDefaultsFalse<false, A, B>              -> B
// IfDefaultsFalse<true, A, B>               -> A
// IfDefaultsFalse<undefined, A, B>          -> B
// IfDefaultsFalse<boolean, A, B>            -> A | B
```

Reply:
Aha, thank you for the explanation. It didn't occur to me that `boolean` would be treated as a union here, though it makes sense that it would.